modules/gapi/src/compiler/gstreaming.cpp

```diff
@@ -123,18 +123,73 @@ bool cv::GStreamingCompiled::pull(cv::GRunArgsP &&outs)
     return m_priv->pull(std::move(outs));
 }
 
-std::tuple<bool, cv::GRunArgs> cv::GStreamingCompiled::pull()
+void cv::detail::constructOptGraphOutputs(const cv::GTypesInfo &out_info,
+                                                cv::GOptRunArgs      &args,
+                                                cv::GOptRunArgsP     &outs)
 {
+    for (auto&& info : out_info)
+    {
+        switch (info.shape)
+        {
+            case cv::GShape::GMAT:
+            {
+                args.emplace_back(cv::optional<cv::Mat>{});
+                outs.emplace_back(&cv::util::get<cv::optional<cv::Mat>>(args.back()));
+                break;
+            }
+            case cv::GShape::GSCALAR:
+            {
+                args.emplace_back(cv::optional<cv::Scalar>{});
+                outs.emplace_back(&cv::util::get<cv::optional<cv::Scalar>>(args.back()));
+                break;
+            }
+            case cv::GShape::GARRAY:
+            {
+                VectorRef ref;
+                util::get<ConstructVec>(info.ctor)(ref);
+                args.emplace_back(cv::util::make_optional(ref));
+                outs.emplace_back(wrap_opt_arg(cv::util::get<optional<VectorRef>>(args.back())));
+                break;
+            }
+            case cv::GShape::GOPAQUE:
+            {
+                OpaqueRef ref;
+                util::get<ConstructOpaque>(info.ctor)(ref);
+                args.emplace_back(cv::util::make_optional(ref));
+                outs.emplace_back(wrap_opt_arg(cv::util::get<optional<VectorRef>>(args.back())));
+                break;
+            }
+            default:
+                cv::util::throw_error(std::logic_error("Unsupported optional output shape for Python"));
+        }
+    }
+}
+
+std::tuple<bool, cv::util::variant<cv::GRunArgs, cv::GOptRunArgs>> cv::GStreamingCompiled::pull()
+{
+    const auto& out_info = m_priv->outInfo();
+    bool is_over = false;
+
+    using runArgs = cv::util::variant<cv::GRunArgs, cv::GOptRunArgs>;
+    if (m_priv->isDesync()) {
+        GOptRunArgs opt_run_args;
+        GOptRunArgsP opt_outs;
+        opt_outs.reserve(out_info.size());
+        opt_run_args.reserve(out_info.size());
+
+        constructOptGraphOutputs(out_info, opt_run_args, opt_outs);
+        is_over = m_priv->pull(std::move(opt_outs));
+        return std::make_tuple(is_over, runArgs(opt_run_args));
+    }
+
     GRunArgs run_args;
     GRunArgsP outs;
-    const auto& out_info = m_priv->outInfo();
     run_args.reserve(out_info.size());
     outs.reserve(out_info.size());
 
-    cv::detail::constructGraphOutputs(m_priv->outInfo(), run_args, outs);
-
-    bool is_over = m_priv->pull(std::move(outs));
-    return std::make_tuple(is_over, run_args);
+    cv::detail::constructGraphOutputs(out_info, run_args, outs);
+    is_over = m_priv->pull(std::move(outs));
+    return std::make_tuple(is_over, runArgs(run_args));
 }
```

Comment:
This whole file is a thin layer over GStreamingExecutor. All this code is relevant to GStreamingExecutor, not the GStreamingCompiled. Please move this into GStreamingExecutor and avoid any additional details exposure in this file.

Reply:
Refactored.